File: [0019][Medium][Remove_Nth_Node_From_End_of_List][LinkedList_2_pointers]/Remove_Nth_Node_From_End_of_List.java
import java.util.List;
import java.util.Arrays;
import java.util.ArrayList;
import java.util.LinkedList;

//Definition for singly-linked list.
public class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }
}

// solution 1, loop twice
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        int length  = 0;
        ListNode first = head;
        while (first != null) {
            length++;
            first = first.next;
        }
        length -= n;
        first = dummy;
        while (length > 0) {
            length--;
            first = first.next;
        }
        first.next = first.next.next;
        return dummy.next;
    }
}

File: [0019][Medium][Remove_Nth_Node_From_End_of_List][LinkedList_2_pointers]/Remove_Nth_Node_From_End_of_List.py
#Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

#solution 2, 2 pointers, once loop
class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        dummyNode = ListNode(None)
        dummyNode.next = head
        first,slow = dummyNode,dummyNode
        for i in range(n):
            first = first.next
        while first.next != None:
            first = first.next
            slow = slow.next
        slow.next = slow.next.next
        return dummyNode.next